hypothesis-python/src/hypothesis/internal/conjecture/shrinking/dfas.py

```diff
@@ -97,26 +97,32 @@ def learn_a_new_dfa(runner, u, v, predicate):
     # just one tricky bit of control flow we're struggling to
     # reduce inside a strategy somewhere and the rest of the
     # test function reduces fine.
-    i = 0
-    while u[i] == v[i]:
-        i += 1
-    prefix = u[:i]
-    assert u.startswith(prefix)
-    assert v.startswith(prefix)
-
-    i = 1
-    while u[-i] == v[-i]:
-        i += 1
-
-    suffix = u[len(u) + 1 - i :]
-    assert u.endswith(suffix)
-    assert v.endswith(suffix)
-
-    u_core = u[len(prefix) : len(u) - len(suffix)]
-    v_core = v[len(prefix) : len(v) - len(suffix)]
-
-    assert u == prefix + u_core + suffix
-    assert v == prefix + v_core + suffix
+    if v.endswith(u):
+        prefix = b""
+        suffix = u
+        u_core = b""
+        v_core = v[: -len(u)]
```

Comment:
Minor: This helps show that it's always safe to index by `-len(u)`.

```suggestion
        assert len(u) > 0
        v_core = v[: -len(u)]
```